{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7edf4686",
   "metadata": {},
   "source": [
    "# Day5 exercise 2\n",
    "\n",
    "- 18.03.2022\n",
    "- Kriti Amin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21a4aaa4",
   "metadata": {},
   "source": [
    "# SQLAlchemy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62795143",
   "metadata": {},
   "source": [
    "SQLAlchemy is the Python SQL toolkit and Object Relational Mapper\n",
    "\n",
    "* [website](https://www.sqlalchemy.org/)\n",
    "* [documentation](https://docs.sqlalchemy.org/en/14/)\n",
    "* [Introduction as Jupyter Notebook](sqlalchemy_into.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05efbcbb",
   "metadata": {},
   "source": [
    "If needed execute following SQLs in MySQL CLI\n",
    "\n",
    "```sql\n",
    "DROP DATABASE IF EXISTS biodb;\n",
    "CREATE DATABASE biodb;\n",
    "SHOW DATABASES like 'biodb';\n",
    "CREATE USER IF NOT EXISTS 'biodb_user'@'localhost' IDENTIFIED BY 'biodb_password';\n",
    "SELECT User FROM mysql.user WHERE User LIKE 'biodb_user';\n",
    "GRANT ALL ON `biodb`.* TO 'biodb_user'@'localhost';\n",
    "FLUSH PRIVILEGES;\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ea5b5d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "from sqlalchemy.orm import Session\n",
    "engine = create_engine('mysql+pymysql://biodb_user:biodb_password@localhost/biodb')\n",
    "session = Session(engine)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96117e86",
   "metadata": {},
   "source": [
    "## Import SQL dump\n",
    "\n",
    "If link to dump file is not provided until now by Christian, please complain!\n",
    "\n",
    "1. Download zip file\n",
    "2. Unzip file\n",
    "\n",
    "execute on CMD/terminal:\n",
    "```bash\n",
    "mysql -u biodb_user -pbiodb_password biodb < C:\\Users\\kriti\\Dropbox\\PC\\Desktop\\BioDB\\biodb.sql\n",
    "```\n",
    "\n",
    "\"C:\\Users\\kriti\\Dropbox\\PC\\Desktop\\BioDB\\biodb.sql\"\n",
    "\n",
    "Should be finished < 1 min."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a03c217",
   "metadata": {},
   "source": [
    "## Model definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "789c07d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"UniProt RDBMS model definition.\"\"\"\n",
    "from sqlalchemy.orm import relationship\n",
    "from sqlalchemy.ext.declarative import declarative_base\n",
    "from sqlalchemy import Column, Integer, String, Table, Text, ForeignKey\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "Base = declarative_base()\n",
    "\n",
    "uniprot__uniprot_keyword = Table('uniprot__uniprot_keyword', Base.metadata,\n",
    "                                 Column('uniprot_id', Integer, ForeignKey('uniprot.id')),\n",
    "                                 Column('uniprot_keyword_id', Integer,\n",
    "                                        ForeignKey('uniprot_keyword.keywordid'))\n",
    "                                 )\n",
    "\n",
    "uniprot__uniprot_host = Table('uniprot__uniprot_host', Base.metadata,\n",
    "                              Column('uniprot_id', Integer, ForeignKey('uniprot.id')),\n",
    "                              Column('uniprot_organism_id', Integer, ForeignKey('uniprot_organism.taxid'))\n",
    "                              )\n",
    "\n",
    "uniprot__uniprot_xref = Table('uniprot__uniprot_xref', Base.metadata,\n",
    "                              Column('uniprot_id', Integer, ForeignKey('uniprot.id')),\n",
    "                              Column('uniprot_xref_id', Integer, ForeignKey('uniprot_xref.id'))\n",
    "                              )\n",
    "\n",
    "uniprot__uniprot_subcellular_location = Table('uniprot__uniprot_subcellular_location', Base.metadata,\n",
    "                                              Column('uniprot_id', Integer, ForeignKey('uniprot.id')),\n",
    "                                              Column('uniprot_subcellular_location_id',\n",
    "                                                     Integer, ForeignKey('uniprot_subcellular_location.id'))\n",
    "                                              )\n",
    "\n",
    "\n",
    "class Uniprot(Base):\n",
    "    \"\"\"Class definition for the UniProt table.\"\"\"\n",
    "\n",
    "    __tablename__ = 'uniprot'\n",
    "\n",
    "    id = Column(Integer, primary_key=True)\n",
    "\n",
    "    accession = Column(String(20), unique=True)\n",
    "    name = Column(String(100), nullable=False, unique=True)\n",
    "    recommended_name = Column(String(255), nullable=True)\n",
    "\n",
    "    taxid = Column(Integer, ForeignKey('uniprot_organism.taxid'), nullable=False, index=True)\n",
    "    organism = relationship(\"Organism\")\n",
    "\n",
    "    function_id = Column(Integer, ForeignKey('uniprot_function.id'), nullable=True)\n",
    "    function = relationship(\"Function\")\n",
    "\n",
    "    gene_names = relationship(\"Gene\", back_populates=\"uniprot\")\n",
    "\n",
    "    gene_symbol = relationship(\"GeneSymbol\", uselist=False, back_populates=\"uniprot\")\n",
    "\n",
    "    keywords = relationship(\n",
    "        \"Keyword\",\n",
    "        secondary=uniprot__uniprot_keyword,\n",
    "        back_populates=\"uniprots\")\n",
    "\n",
    "    xrefs = relationship(\n",
    "        \"Xref\",\n",
    "        secondary=uniprot__uniprot_xref,\n",
    "        back_populates=\"uniprots\"\n",
    "    )\n",
    "\n",
    "    subcellular_locations = relationship(\n",
    "        \"SubcellularLocation\",\n",
    "        secondary=uniprot__uniprot_subcellular_location,\n",
    "        back_populates=\"uniprots\"\n",
    "    )\n",
    "\n",
    "    def __repr__(self):\n",
    "        return self.name\n",
    "\n",
    "    def as_dict(self):\n",
    "        \"\"\"Convert object values to dictionary.\"\"\"\n",
    "        xrefs_grouped = defaultdict(list)\n",
    "        for xref in self.xrefs:\n",
    "            xrefs_grouped[xref.db].append(xref.identifier)\n",
    "        xrefs_grouped = {k: sorted(v) for k, v in xrefs_grouped.items()}\n",
    "\n",
    "        return {\n",
    "            'name': self.name,\n",
    "            'accession': self.accession,\n",
    "            'recommended_name': self.recommended_name,\n",
    "            'taxid': self.taxid,\n",
    "            'function_description': self.function.description if self.function else self.function,\n",
    "            'gene_names': [x.name for x in self.gene_names],\n",
    "            'gene_symbol': self.gene_symbol.symbol if self.gene_symbol else self.gene_symbol,\n",
    "            'keywords': [{'keyword': x.keyword_name, 'id': x.keywordid} for x in self.keywords],\n",
    "            'xrefs': xrefs_grouped,\n",
    "            'subcellular_locations': [x.name for x in self.subcellular_locations],\n",
    "            'organism': self.organism.scientific_name\n",
    "        }\n",
    "\n",
    "\n",
    "class GeneSymbol(Base):\n",
    "    \"\"\"Class definition for the uniprot_gene_symbol table.\"\"\"\n",
    "    __tablename__ = 'uniprot_gene_symbol'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    symbol = Column(String(100), nullable=False, index=True)\n",
    "    uniprot_id = Column(Integer, ForeignKey('uniprot.id'))\n",
    "    uniprot = relationship(\"Uniprot\", back_populates=\"gene_symbol\")\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\"Define repr.\"\"\"\n",
    "        return self.symbol\n",
    "\n",
    "\n",
    "class Gene(Base):\n",
    "    \"\"\"Class definition for the uniprot_gene table.\"\"\"\n",
    "    __tablename__ = 'uniprot_gene'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    name = Column(String(100), nullable=False, index=True)\n",
    "    uniprot_id = Column(Integer, ForeignKey('uniprot.id'))\n",
    "    uniprot = relationship(\"Uniprot\", back_populates=\"gene_names\")\n",
    "\n",
    "\n",
    "class Keyword(Base):\n",
    "    \"\"\"Class definition for the uniprot_keyword table.\"\"\"\n",
    "    __tablename__ = 'uniprot_keyword'\n",
    "    keywordid = Column(Integer, primary_key=True)\n",
    "    keyword_name = Column(String(100), index=True)\n",
    "\n",
    "    uniprots = relationship(\n",
    "        \"Uniprot\",\n",
    "        secondary=uniprot__uniprot_keyword,\n",
    "        back_populates=\"keywords\")\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\"Define repr.\"\"\"\n",
    "        return f\"{self.keyword_name}[{self.keywordid}]\"\n",
    "\n",
    "\n",
    "class Organism(Base):\n",
    "    \"\"\"Class definition for the uniprot_organism table.\"\"\"\n",
    "    __tablename__ = 'uniprot_organism'\n",
    "    taxid = Column(Integer, primary_key=True)\n",
    "    scientific_name = Column(String(255))\n",
    "\n",
    "\n",
    "class SubcellularLocation(Base):\n",
    "    \"\"\"Class definition for the uniprot_subcellular_location table.\"\"\"\n",
    "    __tablename__ = 'uniprot_subcellular_location'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    name = Column(String(100), index=True)\n",
    "    uniprots = relationship(\n",
    "        \"Uniprot\",\n",
    "        secondary=uniprot__uniprot_subcellular_location,\n",
    "        back_populates=\"subcellular_locations\")\n",
    "\n",
    "\n",
    "class Xref(Base):\n",
    "    \"\"\"Class definition for the uniprot_xref table.\"\"\"\n",
    "    __tablename__ = 'uniprot_xref'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    db = Column(String(50), index=True)\n",
    "    identifier = Column(String(100), index=True)\n",
    "    uniprots = relationship(\n",
    "        \"Uniprot\",\n",
    "        secondary=uniprot__uniprot_xref,\n",
    "        back_populates=\"xrefs\")\n",
    "\n",
    "\n",
    "class Function(Base):\n",
    "    \"\"\"Class definition for the uniprot_function table.\"\"\"\n",
    "    __tablename__ = 'uniprot_function'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    description = Column(Text)\n",
    "    uniprots = relationship(\"Uniprot\", back_populates=\"function\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6688cec6",
   "metadata": {},
   "source": [
    "## Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a98e7211",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2A5D_HUMAN,\n",
       " 2A5E_HUMAN,\n",
       " 2A5G_HUMAN,\n",
       " 5HT1B_HUMAN,\n",
       " 5HT1D_HUMAN,\n",
       " A16L1_HUMAN,\n",
       " A16L2_HUMAN,\n",
       " A26L1_HUMAN,\n",
       " AA2AR_HUMAN,\n",
       " AA2BR_HUMAN]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Uniprot).limit(10).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c91abddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = session.query(Uniprot).filter_by(accession=\"P20138\").limit(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b5351e6",
   "metadata": {},
   "source": [
    "print the SQL statement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0da557c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT uniprot.id, uniprot.accession, uniprot.name, uniprot.recommended_name, uniprot.taxid, uniprot.function_id \n",
      "FROM uniprot \n",
      "WHERE uniprot.accession = %s \n",
      " LIMIT %s\n"
     ]
    }
   ],
   "source": [
    "from sqlalchemy.dialects import mysql\n",
    "print(str(query.statement.compile(dialect=mysql.dialect())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e59e3616",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[CD33_HUMAN]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query.all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7d9c93d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd33 = query.one()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3b80feca",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'CD33_HUMAN',\n",
       " 'accession': 'P20138',\n",
       " 'recommended_name': 'Myeloid cell surface antigen CD33',\n",
       " 'taxid': 9606,\n",
       " 'function_description': 'Sialic-acid-binding immunoglobulin-like lectin (Siglec) that plays a role in mediating cell-cell interactions and in maintaining immune cells in a resting state (PubMed:10611343, PubMed:15597323, PubMed:11320212). Preferentially recognizes and binds alpha-2,3- and more avidly alpha-2,6-linked sialic acid-bearing glycans (PubMed:7718872). Upon engagement of ligands such as C1q or syalylated glycoproteins, two immunoreceptor tyrosine-based inhibitory motifs (ITIMs) located in CD33 cytoplasmic tail are phosphorylated by Src-like kinases such as LCK (PubMed:28325905, PubMed:10887109). These phosphorylations provide docking sites for the recruitment and activation of protein-tyrosine phosphatases PTPN6/SHP-1 and PTPN11/SHP-2 (PubMed:10556798, PubMed:10206955, PubMed:10887109). In turn, these phosphatases regulate downstream pathways through dephosphorylation of signaling molecules (PubMed:10206955, PubMed:10887109). One of the repressive effect of CD33 on monocyte activation requires phosphoinositide 3-kinase/PI3K (PubMed:15597323).',\n",
       " 'gene_names': ['CD33', 'SIGLEC3'],\n",
       " 'gene_symbol': 'CD33',\n",
       " 'keywords': [{'keyword': '3D-structure', 'id': 2},\n",
       "  {'keyword': 'Alternative splicing', 'id': 25},\n",
       "  {'keyword': 'Cell adhesion', 'id': 130},\n",
       "  {'keyword': 'Cell membrane', 'id': 1003},\n",
       "  {'keyword': 'Disulfide bond', 'id': 1015},\n",
       "  {'keyword': 'Glycoprotein', 'id': 325},\n",
       "  {'keyword': 'Immunoglobulin domain', 'id': 393},\n",
       "  {'keyword': 'Lectin', 'id': 430},\n",
       "  {'keyword': 'Membrane', 'id': 472},\n",
       "  {'keyword': 'Peroxisome', 'id': 576},\n",
       "  {'keyword': 'Phosphoprotein', 'id': 597},\n",
       "  {'keyword': 'Reference proteome', 'id': 1185},\n",
       "  {'keyword': 'Repeat', 'id': 677},\n",
       "  {'keyword': 'Signal', 'id': 732},\n",
       "  {'keyword': 'Transmembrane', 'id': 812},\n",
       "  {'keyword': 'Transmembrane helix', 'id': 1133}],\n",
       " 'xrefs': {'EMBL': ['AC063977',\n",
       "   'AK304810',\n",
       "   'AY040541',\n",
       "   'AY162464',\n",
       "   'BC028152',\n",
       "   'CH471135',\n",
       "   'M23197'],\n",
       "  'CCDS': ['CCDS33084.1', 'CCDS46157.1', 'CCDS54299.1'],\n",
       "  'PIR': ['A30521'],\n",
       "  'RefSeq': ['NP_001076087.1',\n",
       "   'NP_001171079.1',\n",
       "   'NP_001763.3',\n",
       "   'XP_011525834.1'],\n",
       "  'PDB': ['5IHB', '5J06', '5J0B', '6D48', '6D49', '6D4A', '6TL8', '7AW6'],\n",
       "  'PDBsum': ['5IHB', '5J06', '5J0B', '6D48', '6D49', '6D4A', '6TL8', '7AW6'],\n",
       "  'SMR': ['P20138'],\n",
       "  'BioGRID': ['107383'],\n",
       "  'IntAct': ['P20138'],\n",
       "  'MINT': ['P20138'],\n",
       "  'STRING': ['9606.ENSP00000262262'],\n",
       "  'ChEMBL': ['CHEMBL1842'],\n",
       "  'DrugBank': ['DB00056', 'DB06318'],\n",
       "  'DrugCentral': ['P20138'],\n",
       "  'GuidetoPHARMACOLOGY': ['2601'],\n",
       "  'UniLectin': ['P20138'],\n",
       "  'GlyGen': ['P20138'],\n",
       "  'iPTMnet': ['P20138'],\n",
       "  'PhosphoSitePlus': ['P20138'],\n",
       "  'BioMuta': ['CD33'],\n",
       "  'DMDM': ['116241290'],\n",
       "  'EPD': ['P20138'],\n",
       "  'jPOST': ['P20138'],\n",
       "  'MassIVE': ['P20138'],\n",
       "  'PaxDb': ['P20138'],\n",
       "  'PeptideAtlas': ['P20138'],\n",
       "  'PRIDE': ['P20138'],\n",
       "  'ProteomicsDB': ['30518', '53725', '9878'],\n",
       "  'ABCD': ['P20138'],\n",
       "  'Antibodypedia': ['4419'],\n",
       "  'CPTC': ['P20138'],\n",
       "  'DNASU': ['945'],\n",
       "  'Ensembl': ['ENST00000262262', 'ENST00000391796', 'ENST00000421133'],\n",
       "  'GeneID': ['945'],\n",
       "  'KEGG': ['hsa:945'],\n",
       "  'UCSC': ['uc010eos.2'],\n",
       "  'CTD': ['945'],\n",
       "  'DisGeNET': ['945'],\n",
       "  'GeneCards': ['CD33'],\n",
       "  'HGNC': ['HGNC:1659'],\n",
       "  'HPA': ['ENSG00000105383'],\n",
       "  'MIM': ['159590'],\n",
       "  'neXtProt': ['NX_P20138'],\n",
       "  'OpenTargets': ['ENSG00000105383'],\n",
       "  'PharmGKB': ['PA26210'],\n",
       "  'VEuPathDB': ['HostDB:ENSG00000105383.14'],\n",
       "  'eggNOG': ['ENOG502S41V'],\n",
       "  'GeneTree': ['ENSGT01030000234562'],\n",
       "  'HOGENOM': ['CLU_102212_0_0_1'],\n",
       "  'InParanoid': ['P20138'],\n",
       "  'OMA': ['EGAIVSW'],\n",
       "  'PhylomeDB': ['P20138'],\n",
       "  'TreeFam': ['TF332441'],\n",
       "  'PathwayCommons': ['P20138'],\n",
       "  'Reactome': ['R-HSA-198933', 'R-HSA-6798695'],\n",
       "  'SIGNOR': ['P20138'],\n",
       "  'BioGRID-ORCS': ['945'],\n",
       "  'ChiTaRS': ['CD33'],\n",
       "  'GenomeRNAi': ['945'],\n",
       "  'Pharos': ['P20138'],\n",
       "  'PRO': ['PR:P20138'],\n",
       "  'Proteomes': ['UP000005640'],\n",
       "  'RNAct': ['P20138'],\n",
       "  'Bgee': ['ENSG00000105383'],\n",
       "  'ExpressionAtlas': ['P20138'],\n",
       "  'Genevisible': ['P20138'],\n",
       "  'GO': ['GO:0002765',\n",
       "   'GO:0005654',\n",
       "   'GO:0005777',\n",
       "   'GO:0005794',\n",
       "   'GO:0005886',\n",
       "   'GO:0005887',\n",
       "   'GO:0007155',\n",
       "   'GO:0007165',\n",
       "   'GO:0007267',\n",
       "   'GO:0008285',\n",
       "   'GO:0009897',\n",
       "   'GO:0009986',\n",
       "   'GO:0019903',\n",
       "   'GO:0030246',\n",
       "   'GO:0032691',\n",
       "   'GO:0032717',\n",
       "   'GO:0032720',\n",
       "   'GO:0033691',\n",
       "   'GO:0035579',\n",
       "   'GO:0038023',\n",
       "   'GO:0043312',\n",
       "   'GO:0050714',\n",
       "   'GO:0050776',\n",
       "   'GO:0051926',\n",
       "   'GO:0070821',\n",
       "   'GO:0098609',\n",
       "   'GO:0150102',\n",
       "   'GO:1903615'],\n",
       "  'Gene3D': ['2.60.40.10'],\n",
       "  'InterPro': ['IPR003599',\n",
       "   'IPR007110',\n",
       "   'IPR013106',\n",
       "   'IPR013151',\n",
       "   'IPR013783',\n",
       "   'IPR036179'],\n",
       "  'Pfam': ['PF00047', 'PF07686'],\n",
       "  'SMART': ['SM00409'],\n",
       "  'SUPFAM': ['SSF48726'],\n",
       "  'PROSITE': ['PS50835']},\n",
       " 'subcellular_locations': ['Cell membrane', 'Peroxisome'],\n",
       " 'organism': 'Homo sapiens'}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd33.as_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dc6a511b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Sialic-acid-binding immunoglobulin-like lectin (Siglec) that plays a role in mediating cell-cell interactions and in maintaining immune cells in a resting state (PubMed:10611343, PubMed:15597323, PubMed:11320212). Preferentially recognizes and binds alpha-2,3- and more avidly alpha-2,6-linked sialic acid-bearing glycans (PubMed:7718872). Upon engagement of ligands such as C1q or syalylated glycoproteins, two immunoreceptor tyrosine-based inhibitory motifs (ITIMs) located in CD33 cytoplasmic tail are phosphorylated by Src-like kinases such as LCK (PubMed:28325905, PubMed:10887109). These phosphorylations provide docking sites for the recruitment and activation of protein-tyrosine phosphatases PTPN6/SHP-1 and PTPN11/SHP-2 (PubMed:10556798, PubMed:10206955, PubMed:10887109). In turn, these phosphatases regulate downstream pathways through dephosphorylation of signaling molecules (PubMed:10206955, PubMed:10887109). One of the repressive effect of CD33 on monocyte activation requires phosphoinositide 3-kinase/PI3K (PubMed:15597323).'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd33.function.description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0a7fa2cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CD33', 'SIGLEC3']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.name for x in cd33.gene_names]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d7a10e3",
   "metadata": {},
   "source": [
    "## Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f16f6d45",
   "metadata": {},
   "source": [
    "#### 1. Print the first 3 uniprot entries ordered by uniprot.id and print accession and name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9a99aa17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Q14738', '2A5D_HUMAN')\n",
      "('Q16537', '2A5E_HUMAN')\n",
      "('Q13362', '2A5G_HUMAN')\n"
     ]
    }
   ],
   "source": [
    "query = session.query(Uniprot.accession, Uniprot.name).order_by(Uniprot.id).limit(3).all()\n",
    "for i in query:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c58e972",
   "metadata": {},
   "source": [
    "#### 2. Get entries 100-103 from uniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60cdf499",
   "metadata": {},
   "source": [
    "https://docs.sqlalchemy.org/en/14/orm/tutorial.html#querying"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "afc0807d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[ARIP4_HUMAN, ARMX3_HUMAN, ARMX4_HUMAN]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Uniprot)[100:103]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca5e957",
   "metadata": {},
   "source": [
    "#### 3. Get UniProt function description for name='TREM2_HUMAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a21c020d",
   "metadata": {},
   "outputs": [],
   "source": [
    "trem2 = session.query(Uniprot).filter_by(name='TREM2_HUMAN').one()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "047b930c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Forms a receptor signaling complex with TYROBP which mediates signaling and cell activation following ligand binding (PubMed:10799849). Acts as a receptor for amyloid-beta protein 42, a cleavage product of the amyloid-beta precursor protein APP, and mediates its uptake and degradation by microglia (PubMed:27477018, PubMed:29518356). Binding to amyloid-beta 42 mediates microglial activation, proliferation, migration, apoptosis and expression of pro-inflammatory cytokines, such as IL6R and CCL3, and the anti-inflammatory cytokine ARG1 (By similarity). Acts as a receptor for lipoprotein particles such as LDL, VLDL, and HDL and for apolipoproteins such as APOA1, APOA2, APOB, APOE, APOE2, APOE3, APOE4, and CLU and enhances their uptake in microglia (PubMed:27477018). Binds phospholipids (preferably anionic lipids) such as phosphatidylserine, phosphatidylethanolamine, phosphatidylglycerol and sphingomyelin (PubMed:29794134). Regulates microglial proliferation by acting as an upstream regulator of the Wnt/beta-catenin signaling cascade (By similarity). Required for microglial phagocytosis of apoptotic neurons (PubMed:24990881). Also required for microglial activation and phagocytosis of myelin debris after neuronal injury and of neuronal synapses during synapse elimination in the developing brain (By similarity). Regulates microglial chemotaxis and process outgrowth, and also the microglial response to oxidative stress and lipopolysaccharide (By similarity). It suppresses PI3K and NF-kappa-B signaling in response to lipopolysaccharide; thus promoting phagocytosis, suppressing pro-inflammatory cytokine and nitric oxide production, inhibiting apoptosis and increasing expression of IL10 and TGFB (By similarity). During oxidative stress, it promotes anti-apoptotic NF-kappa-B signaling and ERK signaling (By similarity). Plays a role in microglial MTOR activation and metabolism (By similarity). Regulates age-related changes in microglial numbers (PubMed:29752066). Triggers activation of the immune responses in macrophages and dendritic cells (PubMed:10799849). Mediates cytokine-induced formation of multinucleated giant cells which are formed by the fusion of macrophages (By similarity). In dendritic cells, it mediates up-regulation of chemokine receptor CCR7 and dendritic cell maturation and survival (PubMed:11602640). Involved in the positive regulation of osteoclast differentiation (PubMed:12925681).'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trem2.function.description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ad10560",
   "metadata": {},
   "source": [
    "#### 4. Print accession and gene_symbol for all uniprot entries where ***CD33*** in name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1bb49331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accession :  Q8N5R6 Gene Symbol :  CCDC33\n",
      "Accession :  P20138 Gene Symbol :  CD33\n"
     ]
    }
   ],
   "source": [
    "cd33 = session.query(Uniprot).filter(Uniprot.name.like('%CD33%'))\n",
    "for i in cd33:\n",
    "    print('Accession : ',i.accession, 'Gene Symbol : ', i.gene_symbol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87549434",
   "metadata": {},
   "source": [
    "#### 5. Get all uniprot entries where name in ('CCD33_HUMAN', 'TREM2_HUMAN')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "7eb5bc69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[CCD33_HUMAN, TREM2_HUMAN]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Uniprot).filter(Uniprot.name.in_(['CCD33_HUMAN', 'TREM2_HUMAN'])).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9835b32a",
   "metadata": {},
   "source": [
    "#### 6. Get all uniprot entries where name is *CD33_HUMAN* and accession *P20138*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "916f1c96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[CD33_HUMAN]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Uniprot).filter(Uniprot.name == 'CD33_HUMAN', Uniprot.accession == 'P20138').all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea81f751",
   "metadata": {},
   "source": [
    "#### 7. Get all uniprot entries where name is 'CCD33_HUMAN' or 'TREM2_HUMAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "4bb5b805",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[CD33_HUMAN, TREM2_HUMAN]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import or_\n",
    "session.query(Uniprot).filter(or_(Uniprot.name == 'CD33_HUMAN', Uniprot.name == 'TREM2_HUMAN')).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3c177f8",
   "metadata": {},
   "source": [
    "#### 8. How many functions contains the word Alzheimer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "c8df21c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Function).filter(Function.description.like('%Alzheimer%')).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f42bf606",
   "metadata": {},
   "source": [
    "#### 9. Get all uniprot entries where keyword name is 'Alzheimer disease'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "da4c1340",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[ABCA7_HUMAN,\n",
       " A4_HUMAN,\n",
       " TAU_HUMAN,\n",
       " CATD_HUMAN,\n",
       " APOE_HUMAN,\n",
       " GSK3A_HUMAN,\n",
       " GSK3B_HUMAN,\n",
       " SYUA_HUMAN,\n",
       " DREB_HUMAN,\n",
       " NU2M_HUMAN,\n",
       " PPP5_HUMAN,\n",
       " RGPS2_HUMAN,\n",
       " PSN1_HUMAN,\n",
       " PSN2_HUMAN,\n",
       " ADA10_HUMAN,\n",
       " NU1M_HUMAN,\n",
       " UNC5C_HUMAN,\n",
       " SORL_HUMAN]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.query(Uniprot).join(uniprot__uniprot_keyword).join(Keyword).filter(Keyword.keyword_name=='Alzheimer disease').all()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodb_2022",
   "language": "python",
   "name": "biodb_2022"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
